frame/lending/src/models/borrower_data.rs

```diff
@@ -0,0 +1,150 @@
+	fn max_borrow_for_collateral(&self) -> Result<FixedU128, ArithmeticError> {
+		self.collateral_balance_total_value.safe_div(&self.collateral_factor)
+	}
+
+	/// Determines whether the loan should trigger a liquidation, based on the
+	/// [`current_collateral_ratio`].
+	///
+	/// [`current_collateral_ratio`]: BorrowerData::current_collateral_ratio
+	#[inline(always)]
+	pub fn should_liquidate(&self) -> Result<bool, ArithmeticError> {
+		match self.current_collateral_ratio()? {
+			CollateralRatio::Ratio(ratio) => Ok(ratio < *self.collateral_factor),
+			// No liquidation necessary if the borrower's borrow asset balance has no value
+			CollateralRatio::NoBorrowValue => Ok(false),
+		}
+	}
+
+	/// The current collateral to debt ratio for the borrower. See [`CollateralRatio`] for
+	/// more information.
+	#[inline(always)]
+	pub fn current_collateral_ratio(&self) -> Result<CollateralRatio<FixedU128>, ArithmeticError> {
+		if self.borrow_balance_total_value.is_zero() {
+			Ok(CollateralRatio::NoBorrowValue)
+		} else {
+			// REVIEW: What errors can occur with safe_div?
+			let ratio =
+				self.collateral_balance_total_value.safe_div(&self.borrow_balance_total_value)?;
+			Ok(CollateralRatio::Ratio(ratio))
+		}
+	}
+
+	/// The lowest value the collateral ratio can go before the borrower will be warned about soon
+	/// being under-collateralized.
+	///
+	/// Calculated as follows:
+	///
+	/// ```python
+	/// safe_collateral_factor = collateral_factor + (collateral_factor * under_collateralized_warn_percent)
+	/// ```
+	///
+	/// # Example
+	///
+	/// ```rust,ignore
+	/// let collateral_factor = 2;
+	/// let under_collateralized_warn_percent = Percent::from_percent(10); // 10%
+	///
+	/// let safe_collateral_factor = collateral_factor + (collateral_factor * under_collateralized_warn_percent);
+	///                         // = 2 + (2 * 10%)
+	///                         // = 2 + 0.2
+	///                         // = 2.2
+	/// ```
+	///
+	/// With the above values, if the collateral to debt ratio goes below `2.2`, the account will be
+	/// warned about soon being under-collateralized.
+	#[inline(always)]
+	pub fn minimum_safe_collateral_factor(&self) -> Result<MoreThanOneFixedU128, ArithmeticError> {
+		let collateral_factor_warn_percentage =
+			self.collateral_factor.safe_mul(&self.under_collateralized_warn_percent.into());
+		self.collateral_factor.safe_add(&collateral_factor_warn_percentage?)
+	}
+
+	/// Check if a loan is about to go under-collateralized.
+	///
+	/// Checks the [`current_collateral_ratio`] against the [`minimum_safe_collateral_factor`].
+	///
+	/// [`minimum_safe_collateral_factor`]: BorrowerData::minimum_safe_collateral_factor
+	/// [`current_collateral_ratio`]: BorrowerData::current_collateral_ratio
+	#[inline(always)]
+	pub fn should_warn(&self) -> Result<bool, ArithmeticError> {
+		match self.current_collateral_ratio()? {
+			CollateralRatio::Ratio(ratio) => Ok(ratio < self.minimum_safe_collateral_factor()?),
```

Comment:
so you map that back from enum to error, so may be we should not have no 2 value enum?